blobstore/src/main/java/org/jclouds/blobstore/internal/BaseBlobStore.java

```diff
@@ -354,48 +350,22 @@ protected String putMultipartBlob(String container, Blob blob, PutOptions overri
    protected String putMultipartBlob(String container, Blob blob, PutOptions overrides, ListeningExecutorService executor) {
       ArrayList<ListenableFuture<MultipartPart>> parts = new ArrayList<ListenableFuture<MultipartPart>>();
       MultipartUpload mpu = initiateMultipartUpload(container, blob.getMetadata(), overrides);
-      // Cannot slice InputStream Payload since slice and close mutate the
-      // underlying stream.  Also issue synchronous uploads to avoid buffering
-      // arbitrary amounts of data in-memory.
       Payload payload = blob.getPayload();
-      boolean repeatable = blob.getPayload().isRepeatable();
-      if (!repeatable) {
-         payload = Payloads.newInputStreamPayload(new FilterInputStream((InputStream) payload.getRawContent()) {
-            @Override
-            public long skip(long offset) throws IOException {
-               // intentionally not implemented
-               return offset;
-            }
-
-            @Override
-            public void close() throws IOException {
-               // intentionally not implemented
-            }
-         });
-      }
-
-      try {
-         long contentLength = blob.getMetadata().getContentMetadata().getContentLength();
-         // TODO: inject MultipartUploadSlicingAlgorithm to override default part size
-         MultipartUploadSlicingAlgorithm algorithm = new MultipartUploadSlicingAlgorithm(
-               getMinimumMultipartPartSize(), getMaximumMultipartPartSize(), getMaximumNumberOfParts());
-         long partSize = algorithm.calculateChunkSize(contentLength);
-         int partNumber = 1;
-         while (partNumber <= algorithm.getParts()) {
-            Payload slice = slicer.slice(payload, algorithm.getCopied(), partSize);
+      boolean repeatable = payload.isRepeatable();
+      int partSize = (int) (getMinimumMultipartPartSize() + getMaximumMultipartPartSize()) / 2;
+      int partNumber = 1;
+      int read;
+
+      try (InputStream is = payload.openStream()) {
+         for (byte[] buffer = new byte[partSize]; (read = is.read(buffer)) > -1;) {
+            Payload slice = Payloads.newByteArrayPayload(read == partSize ? buffer : Arrays.copyOf(buffer, read));
```

Comment:
It has been a while since I looked at this code, but I believe that the previous slicing logic could efficiently create parts from repeatable `Payload`s, minimizing memory use.  This code will buffer the entire payload in-memory which won't work when users upload 5 TB objects.